tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/ItemTests.java

```diff
@@ -0,0 +1,261 @@
+import net.neoforged.neoforge.transfer.resources.ItemResource;
+import net.neoforged.neoforge.transfer.transaction.Transaction;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.Test;
+
+public class ItemTests {
+    @Test
+    public void testStackReference() {
+        // Ensure that Container wrappers will try to mutate the backing stack as much as possible.
+        // In many cases, MC code captures a reference to the ItemStack so we want to edit that stack directly
+        // and not a copy whenever we can. Obviously this can't be perfect, but we try to cover as many cases as possible.
+        SimpleContainer container = new SimpleContainer(new ItemStack(Items.DIAMOND, 2));
+        var containerWrapper = VanillaContainerWrapper.of(container);
+        ItemStack stack = container.getItem(0);
+
+        // Simulate should correctly reset the stack.
+        try (Transaction tx = Transaction.open(null)) {
+            containerWrapper.extract(ItemResource.of(Items.DIAMOND), 2, tx);
+        }
+
+        if (stack != container.getItem(0)) throw new AssertionError("Stack should have stayed the same.");
+
+        // Commit should try to edit the original stack when it is feasible to do so.
+        try (Transaction tx = Transaction.open(null)) {
+            containerWrapper.extract(ItemResource.of(Items.DIAMOND), 1, tx);
+            tx.commit();
+        }
+
+        if (stack != container.getItem(0)) throw new AssertionError("Stack should have stayed the same.");
+
+        // Also edit the stack when the item matches, even when the components and the count change.
+        ItemResource oldResource = ItemResource.of(Items.DIAMOND);
+        ItemResource newResource = oldResource.with(DataComponents.MAX_DAMAGE, 10);
+
+        try (Transaction tx = Transaction.open(null)) {
+            containerWrapper.extract(oldResource, 2, tx);
+            containerWrapper.insert(newResource, 5, tx);
+            tx.commit();
+        }
+
+        if (stack != container.getItem(0)) throw new AssertionError("Stack should have stayed the same.");
+        if (!stackEquals(stack, newResource, 5)) throw new AssertionError("Failed to update stack components or count.");
+    }
+
+    @Test
+    public void testContainerWrappers() {
+        ItemResource emptyBucket = ItemResource.of(Items.BUCKET);
+        TestWorldlyContainer testContainer = new TestWorldlyContainer();
+        checkComparatorOutput(testContainer);
+
+        // Create a few wrappers.
+        var unsidedWrapper = VanillaContainerWrapper.of(testContainer);
+        var downWrapper = new WorldlyContainerWrapper(testContainer, Direction.DOWN);
+        var upWrapper = new WorldlyContainerWrapper(testContainer, Direction.UP);
+
+        // Make sure querying a new wrapper returns the same one.
+        if (VanillaContainerWrapper.of(testContainer) != unsidedWrapper) throw new AssertionError("Wrappers should be ==.");
```

Comment:
Can we remove the period to make this error a little more readable that the emphasis is on `==` and not that it should be equal a period